Python/chapter03/p02_stack_min/miguelHx.py

```diff
@@ -0,0 +1,222 @@
+"""
+Python version 3.7.0
+3.2 - Stack Min
+How would you design a stack which, in addition to
+push and pop, has a function min which returns the minimum element?
+Push, pop and min should all operate in O(1) time.
+"""
+
+import copy
+import unittest
+import sys
+from typing import List
+
+
+class StackNode(object):
+
+    def __init__(self, data):
+        self.data = data
+        self.next = None # next is a pointer to a StackNode object
+        self.running_min = sys.maxsize
+
+class MyStack(object):
+    """
+    Stack data structure implementation.
+    Uses LIFO (last-in first-out) ordering.
+    The most recent item added to the stack is
+    the first removed.  Traversal is top to bottom.
+    """
+
+    def __init__(self):
+        self.top = None # top is a pointer to StackNode object
+        self.size = 0
+        return
+    
+    def pop(self) -> int:
+        """
+        Removes the top item from the stack
+
+        Raises:
+            IndexError: raised when pop is attempted on empty stack
+
+        Returns:
+            int: The data at the top of the stack
+        """
+        if self.top is None:
+            raise IndexError('Stack is Empty.')
+        item = self.top.data
+        self.top = self.top.next
+        self.size -= 1
+        return item
+    
+    def push(self, item: int) -> None:
+        """
+        Adds an item to the top of the stack
+        Args:
+            item (int): data we want at the top of stack
+        """
+        t = StackNode(item)
+        t.next = self.top
+        if self.top is None:
+            t.running_min = item
+            self.top = t
+            self.size += 1
+            # print("Stack Node data = {}, local min = {}".format(item, t.running_min))
+            return
+        if item < self.top.data:
+            t.running_min = item
+        else:
+            t.running_min = self.top.running_min
+        # print("Stack Node data = {}, local min = {}".format(item, t.running_min))
+        self.top = t
+        self.size += 1
+
+    def min(self) -> int:
+        """
+        Returns the element with the lowest value in the stack
+
+        Returns:
+            int: min value in stack
+        """
+        return self.top.running_min
```

Comment:
what happens if `self.top` is `None`?

Reply:
Will get an exception for non existing property on None type.

Resolved in ae680053e33e944620aaef98b9c1d6562f49c6c0